frontend/src/main/java/com/google/tripmeout/frontend/storage/InMemoryPlaceVisitStorage.java

```diff
@@ -0,0 +1,102 @@
+package com.google.tripmeout.frontend.storage;
+
+import com.google.common.collect.HashBasedTable;
+import com.google.common.collect.Table;
+import com.google.common.collect.Tables;
+import com.google.tripmeout.frontend.PlaceVisitModel;
+import com.google.tripmeout.frontend.error.PlaceVisitAlreadyExistsException;
+import com.google.tripmeout.frontend.error.PlaceVisitNotFoundException;
+import com.google.tripmeout.frontend.error.TripNotFoundException;
+import com.google.tripmeout.frontend.storage.PlaceVisitStorage;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class InMemoryPlaceVisitStorage implements PlaceVisitStorage {
+  // <tripId, placeId, PlaceVisitModel>
+  Map<String, Map<String, PlaceVisitModel>> storage = new ConcurrentHashMap<>();
+
+  @Override
+  public void addPlaceVisit(PlaceVisitModel placeVisit) throws PlaceVisitAlreadyExistsException {
+    Map<String, PlaceVisitModel> newPlaceMap = new ConcurrentHashMap<>();
+    newPlaceMap.put(placeVisit.placeId(), placeVisit);
+
+    Map<String, PlaceVisitModel> alreadyThereMap =
+        storage.putIfAbsent(placeVisit.tripId(), newPlaceMap);
```

Comment:
As a minor optimization, use `computeIfAbsent` such that we do not allocate and populate a new map if the given trip id has already been inserted. This also simplifies your overall code structure because `placeMap` _always_ contains the correct, current submap:

```java
Map<String, PlaceVisitModel> placeMap = storage.computeIfAbsent(placeVisit.tripId(), key -> new ConcurrentHashMap<>());
// placeMap is never null here and always points to the correct submap
PlaceVisitModel existingPlaceVisit = placeMap.putIfAbsent(placeVisit.placeId(), placeVisit);
if (existingPlaceVisit != null) {
  throw new ...;
}
```

Note: because the double-level insertion is not atomic (without use of external synchronization), there may be data races when deleting trips/visits. Probably not a big deal here--it depends how consistent your API needs to be. For example, this method (`addPlaceVisit`) may return successfully after a successful call to `removeTripPlaceVisits`, in which case the underlying storage no longer holds the submap we just updated here. You'll need to decide if such scenarios are acceptable.